feat: extend conditionNames

GeorgeTaveras1231 opened the pull request:
<!--
  HOLY CRAP a Pull Request. We ❤️ those!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Please place an x (no spaces!) in all [ ] that apply
-->

This PR contains a:

- [ ] **bugfix**
- [x] new **feature**
- [ ] **code refactor**
- [ ] **test update** <!-- if bug or feature is checked, this should be too -->
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

<!--
  Please explain the motivation or use-case for your change.
  What existing problem does the PR solve?
  If this PR addresses an issue, please link to the issue.
-->
This allows consumers to set their own considitionNames and have sass pick this up.

Examples:

Supporting a single entrypoint but targeting multiple themes
```scss
@use 'my-design-tokens';

.class {
  color: my-design-tokens.$color-1
}
```

package.json
```json
{
  "name": "my-design-tokens",
  "exports": {
      ".": {
        "sass": {
          "theme1": "./path-to-theme1.scss",
          "theme2": "./path-to-theme2.scss"
          }
      }
   }
}
```
Webpack config
```
module.exports = {
  resolve: {
    conditionNames: [someFlag ? "theme1" : "theme2", "..."]
  }
}
```

### Breaking Changes

<!--
  If this PR introduces a breaking change, please describe the impact and a
  migration path for existing applications.
-->

### Additional Info

GeorgeTaveras1231 commented:
@alexander-akait @ersachin3112 Looking for comments on this.

This seems like a reasonable extension to the sass loader. I was concerned that this may cause issues if it inherits webpack's  conditionNames, as it would start resolving `import`, or `module` conditions. However, I tested this locally and it seems like it still ignores `import` and `module` exports. My assumption is this is due to webpack's internal handling of conditions, where it seems its able to detect that the reference is made from sass so it doesn't set the `module` or `import` conditions (I read this [here](https://webpack.js.org/guides/package-exports/#reference-syntax)). This makes me feel as though this change is completely safe. And adds some more value to the `sass-loader`.

alexander-akait reviewed (approved):
I am fine with it, but let's add tests:
1. Custom `conditionName` (as in your example)
2. Case where you can show that we avoid using `import`/`module`/etc

GeorgeTaveras1231 commented on test/helpers/getCodeFromSass.js:
```diff
@@ -56,6 +56,27 @@ async function getCodeFromSass(testId, options) {
     testFolder,
     "node_modules/package-with-exports/style.scss"
   );
+  const pathToSassPackageWithExportsFieldsAndCustomConditionReplacer = () => {
+    if (context.packageExportsCustomConditionTestVariant === 1) {
+      return path.resolve(
+        testFolder,
+        "node_modules/package-with-exports-and-custom-condition/style-1.scss"
+      );
+    }
+
+    if (context.packageExportsCustomConditionTestVariant === 2) {
+      return path.resolve(
+        testFolder,
+        "node_modules/package-with-exports-and-custom-condition/style-2.scss"
+      );
+    }
+
+    console.warn(
+      "Expedted to receive .packageExportsCustomConditionTestVariant to properly resolve stylesheet in sass only compilation. "
+    );
+    return "";
+  };
```
There wasn't a great pattern to write the type of tests that made sense here so I had to introduce some new constructs (such as this `context` param) to tell `getCodeFromSass` how to properly resolve the file that the tests need to compare against (based on the custom condition).

GeorgeTaveras1231 commented:
@alexander-akait I've updated the tests per your request.

I did not explicitly add a test to check that there is no conflict with the `import` and `module` conditions -- instead, I moved the import condition higher up in the package.json that was being used ([here](https://github.com/webpack-contrib/sass-loader/pull/1092/files#diff-552915406e296019a5c9aa962293b9ca2e82ff498cd0b080efe0108abbd425e8) and [here](https://github.com/webpack-contrib/sass-loader/pull/1092/files#diff-b16360c2fac898454b4d558ac0d0fdf3b747ebe2bb4527164e5b4075a356a7e9)). By doing this, we are stress testing this potential issue, because the `import` condition should take precedence (given they come first in the list of exports). Therefore, if there was a conflict, the current tests would fail because they would be unable to resolve the desired scss files.

Let me know if you would like an explicit test but that felt like an overkill in an already expensive testsuite 😅

alexander-akait reviewed (approved):
I am fine with it, there is only one exception, if you will have for example `theme1` and it will point to js file (or json) you will get an error, so I recommend to use `sass:theme1` to avoid such problems (mostly webpack will not have the such problem due `restriction`, but Node.js or other runtimes potential can have)

GeorgeTaveras1231 commented:
Thanks @alexander-akait ! 💪🏽
